### DISCUSSION TITLE
Systems Design and IPC General

### DISCUSSION THREAD

**Discussion by 4c3e** [markqvist/Reticulum#254] [General]
I wanted to open up a discussion page focused on using Reticulum for inter-process communication on some sort of future computer systems. 
@markqvist gave a nice description of what is possible here:

> All of our computing systems live squarely within the terminology of a 1920's filing cabinet. That's literally the mental construction around what a computer is. A mechanistic animation of a filing cabinet, you know, one of those old metal hunks with plenty of drawers filled with "folders" and "files". Yes, that's what our computers are. Except that they're not. Since all computers today are universal turing machines (at least up to their memory limits), they can be made to represent or model any abstract or concrete concept (from the real world, or from the imagination, or a combination). The UNIX philosophy of "everything is a file" is incredibly powerful, but only up to the limits of the conceptual domain it exists within. There's a lot of other things in the world than "files", and our computers have a hard time dealing with them. This is severely limiting our creative capabilities and defining how we think way too much.
> 
> I want a computing system that does not draw the "filing cabinet" conceptual domain around itself. The kind of operating environment I would like to have could be termed something like "metamorphic computing". There is no operating system. There is no single-channel stdio. There is a hyper-efficient kernel providing the hardware abstration layer, and basic system-independent functionality for an operating environment. There is no files. Everything is abstract objects that can contain any sort of data structures and/or code. Any object can interact with, inspect, modify or communicate with any other object (according to rules, of course) on the system itself, or on other systems. Objects decide what kind of structures to materialise, from simple text to geometries, code, audio or anything else. Any object can erect internal representational domains to place its outputs into.
> 
> All abstract objects communicate with each other and the operating environment layers via Reticulum, making it possible to interact with objects on any system, or even render a hybrid environment that runs across thousands of devices (each one doing their share of the processing). This also makes it easy to control what and who can interact with what parts of your own systems. You can think of it as a sort hyper-connected, possibly distributed (when needed), always efficient, easily reconfigurable abstract simulation of any sort of abstract system you can imagine. This is the computer I want.
> 
> The human interaction with all of this is completely abstracted. You can run your entire "computer" on a very simple device, that simulates the metamorphic environment you want to operate in, and run the interaction interface on a powerful GPU-equipped system, rendering everything in 3D. Or interact directly with the operating environment running on the low-powered device over a console describing all the environments and states as textual representations.
> 
> This conecptual construction is insanely powerful and interesting. And it probably makes no sense until I give some more concrete examples of how such systems would look, how they would work, and what you could do with them.

Some useful resources for those interested:
Plan 9 from Bell Labs - http://doc.cat-v.org/plan_9/4th_edition/papers/9
The Organization of Networks in Plan 9 - http://doc.cat-v.org/plan_9/4th_edition/papers/net/
Minix - https://www.youtube.com/watch?v=bx3KuE7UjGA & https://wiki.minix3.org/doku.php?id=www:documentation:start
Project Oberon - https://people.inf.ethz.ch/wirth/ProjectOberon1992.pdf

**Comment by 4c3e** [comment]
Obviously writing an entire OS focused on Reticulum networking is way out of scope right now, but I think one interesting idea Reticulum application developers could focus on right now is to represent processes as objects that can be fully utilized through an interface of Reticulum destinations. 
An example given in a later discussion: 
> you could write a basic windowing system that exposed data about its windows, and events like mouse clicks and keyboard inputs as RNS destinations, and then you could have other software that reads and writes to these destinations, so each individual running window could be anywhere in the RNS mesh.
> Kind of expanding on the Unix idea where “everything is a file”, I’ve just been having a bit of fun thinking about computing systems where “everything is a RNS destination”

**Comment by avon61002** [comment]
A distributed smalltalk, or some sort of [E lang](https://en.wikipedia.org/wiki/E_(programming_language)) inspired actor system is a really good way of approaching this problem imo.

Modern options:
- https://newspeaklanguage.org/
- https://spritely.institute/goblins/
- https://syndicate-lang.org/

this project also looks interesting: https://applied-langua.ge/~gnuxie/posts/utena-introduction.html

I'm working now on building a "proxy actor" system in python on top of [syndicate-py ](https://git.syndicate-lang.org/syndicate-lang/syndicate-py). The idea here is that for passing local messages, syndicate's message passing infrastructure will be used, and then remote objects can be messaged by resolving their reticulum destination into a "proxy" local capability address and then sending normal-looking messages to these proxy address. If the underlying reticulum channel fails for whatever reason, the proxy actor will be destroyed, allowing syndicate's built in error handling systems to be used.  

In a similar vein, implementing reticulum as an [OCapN netlayer](https://github.com/ocapn/ocapn/blob/main/draft-specifications/Netlayers.md) will go a long way towards realizing this vision.

**Comment by LinuxinaBit** [comment]
I've been conceptualizing Java 'objects' as folders containing files that hold data ('variables') and execute code ('methods').
That is all contained in a sort of virtual filesystem in memory during execution, separate from the filesystem containing the code you actually wrote.
'Everything is an object' is just 'everything is a file' but with a different name.

Personally I nominate that we call virtual desktops 'groupings' and windows 'instances'.
Might be hard to explain to normal people though :thinking:

**Reply by LinuxinaBit** [reply]
In all seriousness, https://www.redox-os.org/ looks to be the most useful brand-new OS.
It doesn't throw away useful terminology or ways of thinking, but it fixes things that need to be fixed.

**Reply by LinuxinaBit** [reply]
It's fine to reinvent the wheel, but please, _please_ don't try to reinvent the circle.
https://en.wikipedia.org/wiki/Not_invented_here/

Edit: and _especially_ please don't try to reinvent the computing equivalent of **math itself** like the people behind urbit: https://wejn.org/2021/02/urbit-good-bad-insane/

**Reply by avon61002** [reply]
The main issue of building on top of Unix-like systems is that they are all built upon a role-based authentication model. When I run a solitaire game in my Linux userspace, that solitaire executable has all the same privileges as me, including reading and modifying my files, accessing the network etc. Sure we could throw the solitaire game in a sandbox, but what if we wanted to compose this solitaire game with other processes? For example what if we now wanted to control the solitaire game remotely from a phone? Now we need to properly sandbox the solitaire game on the host device, while "letting out" only the precise network connections we deem appropriate, then create a secondary control application on the mobile device, which will have it's own sandbox etc. The whole system quickly gets complicated, as sandboxing does not compose well.

Now let's imagine that the solitaire game ran on a "reticulum os". When we ran the program, we would give it two destinations, one destination representing the window it will draw to, and the other representing a file it can read and write high scores to. Under this model solitaire is given the least amount of capabilities it needs to run, and nothing more. When it needs to draw to the screen, it sends messages over a link to the window destination (which, because of reticulum, can now be on any arbitrary device on the network!), when it needs to write a new high score it sends packets to the "file" destination. Because reticulum allows us to easily create new destinations, we can freely create and destroy destinations as we please, for example, if we want to revoke the game's access to the high score file, we can wipe the existing destination we have mapped to the file, so any new packets sent to the old destination won't be routed anywhere. Also, it is now impossible for solitaire to access any random personal files on your device, as it does not know the reticulum address to message them. This way of reasoning about systems is called "capability-based security" and there are a number of great resources for learning about these systems: 
- http://www.erights.org/talks/thesis/markm-thesis.pdf (chapter 3)
- https://files.spritely.institute/papers/spritely-core.html#caps-as-programming

You're right that it is not useful to reinvent the wheel, but these systems already exist! If you replace "reticulum os" with "smalltalk" or "scheme" these systems have been around since the 1970's! The graphical user interface you're currently using, along with the traditional computer mouse, and many other personal computing ideas, came out of the Xerox PARC lab in the 70's which primarily used smalltalk as the "os". You could even (facetiously) say our modern linux personal computers are [badly reinventing the smalltalk wheel](https://www.youtube.com/watch?v=uknEhXyZgsg) ;)

For practical purposes I do think that linux is by far the best current option as an "hardware abstraction layer" for hosting a virtual machine or a message-passing daemon like the [syndicate-server](https://git.syndicate-lang.org/syndicate-lang/syndicate-rs/src/branch/main). It also is possible to build a capability-based userspace on top of the linux kernel, this is one aspect of the [Genode project](https://genode.org/). But building an enormous tower of abstractions is usually detrimental to performance, privacy, and security. It also means a single user of such a system could never reasonably understand the entire system as it is far too complex. The STEPS project showed that it is not totally out of the question to implement a personal computing environment using many orders of magnitude less lines of code than the status quo: https://tinlizzie.org/VPRIPapers/tr2012001_steps.pdf. My hope is some day a from-scratch capability-based OS will come along with good hardware support, and become a much better host environment for an efficient message-passing substrate. 

Some other related links:
- [Design Principles Behind Smalltalk](https://www.cs.virginia.edu/~evans/cs655/readings/smalltalk.html)
- [CLOSOS: Specification of a Lisp operating system](http://metamodular.com/closos.pdf)

**Reply by LinuxinaBit** [reply]
This may or may not be related: https://xkcd.com/2044/

**Reply by LinuxinaBit** [reply]
I will say that that is a good direction to move in, but I don't think Reticulum is necessarily the right thing to start with, and I doubt the need for a complete re-invention of the modern operating system.

**Reply by LinuxinaBit** [reply]
> Under this model solitaire is given the least amount of capabilities it needs to run, and nothing more

More practically; who assigns these capabilities?

If you can't trust the user to not run something stupid on their system, it's reasonable to assume you're not going to trust the capabilities they assign.
If you can't trust program developers to not be malicious, there is no way you're going to trust the capabilities _they_ assign either.

So who do you trust, and how do you assign that trust?

**Reply by avon61002** [reply]
> More practically; who assigns these capabilities?

Great question! We have 2 ubiquitous environments that use capability-like systems already: browsers and mobile OS's. Both have permission UIs that force users to manually give the app or website access to certain OS capabilities, so I would imagine the highest-level capabilities given to processes would be done through a UI. For example on first run there would be an OS-level popup saying "allow solitaire to write to a window, and read/write to a specific file?".

The bigger theme of "trust and capabilities" is a really interesting topic, there's a fantastic talk by Mark S Miller that touches on this which I highly recommend: https://www.youtube.com/watch?v=NAfjEnu6R2g

Edit: The tldr is that you generally trust people the same way you trust our current systems but now you have a better foundation (applications on average will have significantly reduced capabilities than they do compared to our current systems). And in the case were you do start running malicious code, capabilities give you the ability to revoke access at a later date, which at least gives you some form of recourse.

**Reply by LinuxinaBit** [reply]
There's a reason our current systems don't prompt more than they do.
The certainly could, but everyone learned from the whole Windows Vista UAC over-prompting situation _real fast_.

Scott Charney, head of Microsoft's Trustworthy Computing division during the Vista era admits as much: https://www.youtube.com/watch?v=77f-C4T4ff0
And it's not just 'normal' users who hate prompts: https://www.youtube.com/watch?v=w33BY38mtL4

Heck, I don't think _I_ could bear being prompted for every access to every thing that every program on my system needs.
It's even backed up by some actual research: https://lawsofux.com/choice-overload/

**Reply by LinuxinaBit** [reply]
Other sources claim that UAC was purposefully _designed_ to over-prompt and annoy users so developers would stop asking for privilege escalation, but think about the implications of that.

Prompting the user with any frequency at all is _so annoying_ that it was used on purpose to make 3rd party developers stop doing things that triggered it, and you're saying you want to have a prompt for essentially _everything_ a program needs to do on your system?

You're basically asking developers to not use your system, and it would badger the heck out of users in the mean time.

**Reply by LinuxinaBit** [reply]
The fact of the matter is that a normal computer user places a _gigantic_ amount of trust in the developers of every program on said computer, and the OS itself for that matter.

There are things we can do to reduce this, but there's a limit to what people will accept.
It is a compromise, there are no perfect solutions.

**Reply by faragher** [reply]
> Let's translate what you want to do into current year terms. Read it line by line for the full user experience:
> 
> ```
> 1. You install Firefox
> 2. You launch Firefox
> 3. Firefox asks for access to share=network for internet access
> 4. You click yes
> 5. Firefox asks for access to share=ipc for Inter-process communications
> 6. You click yes
> 7. Firefox asks for access to socket=x11
> 8. You click yes
> 9. Firefox asks for access to socket=Wayland
> 10. You click yes, even though you aren't running Wayland yet
> 11. Firefox asks for access to socket=fallback-x11
> 12. You click yes, just in case
> 13. Firefox asks for access to socket=pulseaudio
> 14. You click yes
> 15. Firefox asks for access to socket=pcsc
> 16. You click no, you don't have a smart card
> 17. Firefox asks for access to socket=cups
> 18. You click yes
> 19. Firefox asks for access to device=dri
> 20. You click yes, you're getting tired
> 21. Firefox asks for access to ~/.mozilla
> 22. You click yes
> 23. Firefox asks for MOZ_ENABLE_WAYLAND=1
> 24. You click yes
> 25. Firefox asks for MOZ_USE_XINPUT2=1
> 26. You click yes
> 27. Firefox asks for org.freedesktop.NetworkManager
> 28. You click yes
> 29. Firefox asks for org.gnome.SessionManager
> 30. You click yes
> 31. Firefox asks for org.gtk.vfs.*
> 32. You click yes
> 33. Firefox asks for org.freedesktop.ScreenSaver
> 34. You click yes
> 35. Firefox asks for org.a11y.Bus
> 36. You click yes
> 37. Firefox asks for org.freedesktop.FileManager1
> 38. You click yes
> 39. Firefox asks for org.freedesktop.Notifications
> 40. You click yes
> 41. Firefox asks for io.gitlab.librewolf.*
> 42. You click yes
> 43. Firefox asks for org.mpris.MediaPlayer2.firefox.*
> 44. You click yes
> 45. Firefox asks for io.gitlab.firefox.*
> 46. You click yes
> 47. It's finally done.
> 48. Oh, but you want LibreOffice too?
> 49. You install LibreOffice
> 50. You try to open a document
> 51. LibreOffice asks for...
> 52. You go back to whatever you were using before because you have better things to do with your time.
> ```
> 
> The kicker? If this was one prompt, the user would skip by it faster than clicking 'Accept' on the iTunes Terms and Conditions.
> 
> Note: These permissions are the _real_ permissions given to the LibreWolf Flatpak. Be honest, did you notice the suspicious permissions at all? (I don't have the Firefox Flatpak installed right now)

That actually sounds useful. I already have a deny by default firewall, uMatrix blocking all third-party requests, and my default browser is blocked so anything that tries to call home that way gets blocked as well. Is it a hassle sometimes? Yes, but that's the price you pay. Same with hijacking all DNS requests to my local router and blacklisting various sites.

If nothing else, it shows you what people are doing with your system. GForce User Experience calls home once per second until it gets a handshake. That led to some log spam, and then it led to me removing the software and never re-installing it. Otherwise you have random programs calling home at all times for reasons varying from legitimate to malicious. Having to grant permissions for a new program is very little to ask for safety.

**Reply by LinuxinaBit** [reply]
That is the exception, and I do think prompts for things like ‘let program run in background’ are pragmatic.
I just think that for anything more, it is only really beneficial for power users, and that’s where Flatseal and/or more advanced system auditing tools come into play.

**Reply by LinuxinaBit** [reply]
Bringing this back to the main discussion, I think my big problem with this fully abstracted object-oriented operating system idea is that it seems like one of those deceptively simple ideas that are vague and hard to design around in reality.

Maybe that’s just my pre-existing mentality getting in the way of seeing things properly, and I’m sure such a system could do a lot, but I can’t see how these grand ideas of fully hardware-agnostic computing could become a reality without massive amounts of overhead.

Sure, there are problems that need to be solved, but this feels more like one of those 80s concept car drawings rather than the Prius everyone actually needs.

**Reply by LinuxinaBit** [reply]
> When we ran the program, we would give it two destinations, one destination representing the window it will draw to, and the other representing a file it can read and write high scores to. Under this model solitaire is given the least amount of capabilities it needs to run, and nothing more. When it needs to draw to the screen, it sends messages over a link to the window destination (which, because of reticulum, can now be on any arbitrary device on the network!), when it needs to write a new high score it sends packets to the "file" destination.

Some parts of that actually sound a lot like Wayland and [Waypipe](https://github.com/neonkore/waypipe)
Other parts like Flatpak.
People know the problems and they want to solve them.

Things are moving in the right direction, and projects like RedoxOS are making ground-up changes that look fairly interesting.

I do worry that conflating networking and computing as entirely interchangeable misses a lot of nuance, but I think I’ll leave it at that because this discussion is way above my pay-grade.
Thank you and have a great day <3
